include/boost/multiprecision/integer.hpp

```diff
@@ -190,50 +190,75 @@ BOOST_MP_CXX14_CONSTEXPR typename std::enable_if<boost::multiprecision::detail::
 }
 
 template <class Integer>
-BOOST_MP_CXX14_CONSTEXPR typename std::enable_if<boost::multiprecision::detail::is_integral<Integer>::value, Integer>::type sqrt(const Integer& x, Integer& r)
+BOOST_MP_CXX14_CONSTEXPR typename std::enable_if<boost::multiprecision::detail::is_integral<Integer>::value, Integer>::type karatsuba_sqrt(const Integer& x, Integer& r, size_t bits)
 {
-   //
-   // This is slow bit-by-bit integer square root, see for example
-   // http://en.wikipedia.org/wiki/Methods_of_computing_square_roots#Binary_numeral_system_.28base_2.29
-   // There are better methods such as http://hal.inria.fr/docs/00/07/28/54/PDF/RR-3805.pdf
-   // and http://hal.inria.fr/docs/00/07/21/13/PDF/RR-4475.pdf which should be implemented
-   // at some point.
-   //
-   Integer s = 0;
-   if (x == 0)
-   {
-      r = 0;
-      return s;
+#ifndef BOOST_MP_NO_CONSTEXPR_DETECTION
+   // std::sqrt is not constexpr by standard, so use this 
+   if (BOOST_MP_IS_CONST_EVALUATED(bits)) {
+      if (bits <= 4) {
+         if (x == 0) {
+            r = 0;
+            return 0;
+         }
+         else if (x < 4) {
+            r = x - 1;
+            return 1;
+         }
+         else if (x < 9) {
+            r = x - 4;
+            return 2;
+         }
+         else {
+            r = x - 9;
+            return 3;
+         }
+      }
    }
-   int g = msb(x);
-   if (g == 0)
-   {
-      r = 1;
-      return s;
+   else
+#endif
+   // we can calculate it faster with std::sqrt
+   if (bits <= 64) {
+      const uint64_t int32max = uint64_t((std::numeric_limits<uint32_t>::max)());
+      uint64_t val = static_cast<uint64_t>(x);
+      uint64_t s = static_cast<uint64_t>(std::sqrt(static_cast<long double>(val)));
+      // converting to long double can loose some precision, and `sqrt` can give eps error, so we'll fix this
+      // this is needed
+      while (s > int32max || s * s > val) s--;
+      // in my tests this never fired, but theoretically this might be needed
+      while (s < int32max && (s + 1) * (s + 1) <= val) s++;
+      r = Integer(val - s * s);
+      return Integer(s);
    }
-
-   Integer t = 0;
-   r         = x;
-   g /= 2;
-   bit_set(s, g);
-   bit_set(t, 2 * g);
-   r = x - t;
-   --g;
-   do
-   {
-      t = s;
-      t <<= g + 1;
-      bit_set(t, 2 * g);
-      if (t <= r)
-      {
-         bit_set(s, g);
-         r -= t;
-      }
-      --g;
-   } while (g >= 0);
+   // https://hal.inria.fr/file/index/docid/72854/filename/RR-3805.pdf
+   size_t b = bits / 4;
```

Comment:
OK, your friendly curmudgeon here ;)

One of the things we found when implementing Karatsuba multiplication was that variable/temporary control was critical to performance.  I count at least 7 temporaries here, I admit it would make the code less neat, but some variable reuse would be a good thing.  This won't make much difference for fixed-precision cpp_int's, but for variable precision case, it saves a bunch of memory allocations.

The other thing we did - and this is very cpp_int specific - was to figure out up front how much memory would be required for all the variables used in the algorithm (right through all the recursions), and allocate that up front, then share it up between the variables in each recursion.  See https://github.com/boostorg/multiprecision/blob/b725af5aef21b640bb19e99ca3104e0f082cda05/include/boost/multiprecision/cpp_int/multiply.hpp#L228 and following code.

However, I do appreciate this is already a noticeable improvement, so if you've had enough for now, that's fair enough frankly!

Reply:
I also forgot to say.... THANK YOU for this, it's much needed for sure.